mne/annotations.py

```diff
@@ -214,6 +214,46 @@ def __iadd__(self, other):
                                                  other.orig_time))
         return self.append(other.onset, other.duration, other.description)
 
+    def __iter__(self):
+        for idx in range(len(self.onset)):
+            yield (self.onset[idx], self.duration[idx], self.description[idx])
+
+    def __getitem__(self, key):
+        """Propagate indexing and slicing to the underlying numpy structure."""
+        try:
+            if isinstance(key, slice) or isinstance(key, list):
+                print('__getitem__ with slice: {}'.format(key))
+                out = Annotations(onset=self.onset[key],
+                                  duration=self.duration[key],
+                                  description=self.description[key],
+                                  orig_time=self.orig_time)
+            elif isinstance(key, int):
+                print('__getitem__ with integer: {}'.format(key))
+                out = Annotations(onset=[self.onset[key]],
+                                  duration=[self.duration[key]],
+                                  description=[self.description[key]],
+                                  orig_time=self.orig_time)
```

Comment:
I would return a tuple like

(self.onset[key], self.duration[key], self.description[key])

this would be consistent with the iterator.

or it could be a dict with keys onset, duration, description like when you index a pandas dataframe...

thoughts @jona-sassenhagen ?

Reply:
After talking to @agramfort offline he convinced me to return a dictionary not an `Annotations` when indexing with a single integer.

It has the advantage that if we ever want to extend the annotations, with more fields we won't break people's code.

What do you think?

Reply:
Why don't we return an Annotations object for the iterator and for slicing/indexing operations? It has the same extend-ability because whatever we add will be attributes of the Annotations class immediately.

This is also basically what we do with the Epochs class, so it's more consistent with that, too.

Reply:
@mne-tools/contributors ?

Reply:
So the proposals are to return:

- annotations object
- tuple
- dict

?

I haven't worked with annotations enough to have an opinion ...

Reply:
Yep, so concretely:
```
# annot
for a in raw.annotations:
    onset, duration, desc = a.onset[0], a.duration[0], a.desc[0]
    ...
# tuple
for onset, duration, desc in raw.annotations:
    ...
# dict
for a in raw.annotations:
    onset, duration, desc = a['onset'], a['duration'], a['desc']
    ...
```
The only advantage of an `dict` approach over iterating over annotations objects I see is that it avoids the `[0]`, but this does not seem worth introducing more API inconsistency to me.

Reply:
The `annot` will also not break if you start adding a channel name for annotations.

So the question is, do we value internal package consistency (Annotations iterating like Epochs) or consistency with what Pandas does here?

Reply:
I think the argument you are really trying to make has to do with how conceptualize iteration over an N-dimensional object: is what you get in the loop (N-1)D (like arrays, lists, tuples, pandas, etc.) or ND (like epochs). I can see why this would be useful, even if it's less like what we usually do, so I can live with it.

(FWIW, I don't quite see why the iterable argument helps, since it actually holds already for the `epochs` object (`epochs[k] == e for k, e in enumerate(epochs)`) and would hold for the `annotations` API (`annot[k] == a for k, a in annotations`). It seems to actually favor "iter yields Annotations" if `__getitem__` always returns `Annotations`; in order for the relationship to hold for "iter yields dict", we'd need `Annotations | dict` to be returned by `__getitem__`, depending on whether the result is slice vs int...)